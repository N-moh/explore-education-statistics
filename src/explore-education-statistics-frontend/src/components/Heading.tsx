import classNames from 'classnames';
import React, { HTMLAttributes, ReactNode } from 'react';

type HeadingElementProps = {
  children: ReactNode;
} & HTMLAttributes<HTMLHeadingElement>;

type Props = {
  size: 'xl' | 'l' | 'm' | 's';
} & HeadingElementProps;

const headingTags = {
  l: 'h2',
  m: 'h3',
  s: 'h4',
  xl: 'h1',
};

const Heading = ({ children, size, ...props }: Props) =>
  React.createElement<HTMLAttributes<HTMLHeadingElement>>(
    headingTags[size],
    {
      ...props,
      className: classNames(props.className, `govuk-heading-${size}`),
    },
    children,
  );

export default Heading;

export const H1 = ({ children, ...props }: HeadingElementProps) => (
  <Heading {...props} size="xl">
    {children}
  </Heading>
);

export const H2 = ({ children, ...props }: HeadingElementProps) => (
  <Heading {...props} size="l">
    {children}
  </Heading>
);

export const H3 = ({ children, ...props }: HeadingElementProps) => (
  <Heading {...props} size="m">
    {children}
  </Heading>
);

export const H4 = ({ children, ...props }: HeadingElementProps) => (
  <Heading {...props} size="s">
    {children}
  </Heading>
);
