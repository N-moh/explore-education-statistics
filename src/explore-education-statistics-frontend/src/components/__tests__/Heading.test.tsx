import React from 'react';
import { render } from 'react-testing-library';
import Heading, { H1, H2, H3, H4 } from '../Heading';

describe('Heading', () => {
  test('renders `xl` size correctly as h1', () => {
    const { container } = render(<Heading size="xl">Test Heading 1</Heading>);
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders `l` size correctly as h2', () => {
    const { container } = render(<Heading size="l">Test Heading 2</Heading>);
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders `m` size correctly as h3', () => {
    const { container } = render(<Heading size="m">Test Heading 3</Heading>);
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders `s` size correctly as h4', () => {
    const { container } = render(<Heading size="s">Test Heading 4</Heading>);
    expect(container.innerHTML).toMatchSnapshot();
  });
});

describe('H1', () => {
  test('renders correctly as `xl` size', () => {
    const { container } = render(<H1>Test Heading 1</H1>);
    expect(container.innerHTML).toMatchSnapshot();
  });
});

describe('H2', () => {
  test('renders correctly as `l` size', () => {
    const { container } = render(<H2>Test Heading 2</H2>);
    expect(container.innerHTML).toMatchSnapshot();
  });
});

describe('H3', () => {
  test('renders correctly as `m` size', () => {
    const { container } = render(<H3>Test Heading 3</H3>);
    expect(container.innerHTML).toMatchSnapshot();
  });
});

describe('H4', () => {
  test('renders correctly as `xl` size', () => {
    const { container } = render(<H4>Test Heading 4</H4>);
    expect(container.innerHTML).toMatchSnapshot();
  });
});
