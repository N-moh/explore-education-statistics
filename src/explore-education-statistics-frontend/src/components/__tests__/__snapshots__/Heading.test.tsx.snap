// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`H1 renders correctly as \`xl\` size 1`] = `

<h1 class="govuk-heading-xl">
  Test Heading 1
</h1>

`;

exports[`H2 renders correctly as \`l\` size 1`] = `

<h2 class="govuk-heading-l">
  Test Heading 2
</h2>

`;

exports[`H3 renders correctly as \`m\` size 1`] = `

<h3 class="govuk-heading-m">
  Test Heading 3
</h3>

`;

exports[`H4 renders correctly as \`xl\` size 1`] = `

<h4 class="govuk-heading-s">
  Test Heading 4
</h4>

`;

exports[`Heading renders \`l\` size correctly as h2 1`] = `

<h2 class="govuk-heading-l">
  Test Heading 2
</h2>

`;

exports[`Heading renders \`m\` size correctly as h3 1`] = `

<h3 class="govuk-heading-m">
  Test Heading 3
</h3>

`;

exports[`Heading renders \`s\` size correctly as h4 1`] = `

<h4 class="govuk-heading-s">
  Test Heading 4
</h4>

`;

exports[`Heading renders \`xl\` size correctly as h1 1`] = `

<h1 class="govuk-heading-xl">
  Test Heading 1
</h1>

`;
